src/shared/index.ts

```diff
@@ -0,0 +1,220 @@
+// Licensed under the MIT License.
+
+import { ArtifactLocation, Log, Location, Region, Result } from 'sarif';
+
+type JsonLocation = { line: number, column: number } // Unused: pos
+type JsonRange = { value: JsonLocation, valueEnd: JsonLocation } // Unused: key, keyEnd
+export type JsonMap = Record<string, JsonRange>
+
+export type ResultId = [string, number, number]
+type _RegionBytes = [number, number] // byteOffset, byteLength
+type _RegionStartEndLineCol = [number, number, number, number] // start line, start col, end line, end col
+export type _Region
+    = number // single line
+    | _RegionBytes
+    | _RegionStartEndLineCol
+
+// The extended members we're adding here are prefixed with an underscore.
+// Using this marker to easily get a sense of how much we're depending on these extended members.
+// Once this design stabilities, these underscores will likely be removed as
+// I don't see a need for them in the long term (after the design phase).
+declare module 'sarif' {
+    interface Log {
+        _uri: string
+        _uriUpgraded?: string // Only present if upgraded.
+        _jsonMap?: JsonMap // Only used by the "extension" side for navigating original SARIF sources. The "panel" side does not need this feature and thus does not use this field.
+        _augmented: boolean
+        _distinct: Map<string, string> // Technically per Run, practically doesn't matter right now.
+    }
+
+    interface Run {
+        _index: number
+    }
+
+    interface Result {
+        _log: Log
+        _run: Run
+        _id: ResultId
+        _logRegion?: _Region
+        _uri?: string
+        _uriContents?: string // ArtifactContent. Do not use this uri for display.
+        _relativeUri?: string
+        _region?: _Region
+        _line: number // -1 if empty.
+        _rule?: ReportingDescriptor
+        _message: string // '—' if empty.
+        _markdown?: string
+        _suppression?: 'not suppressed' | 'suppressed'
+    }
+}
+
+// console.log(format(`'{0}' was not evaluated for check '{2}' as the analysis is not relevant based on observed metadata: {1}.`, ['x', 'y', 'z']))
+function format(template: string | undefined, args?: string[]) {
+    if (!template) return undefined;
+    if (!args) return template;
+    return template.replace(/{(\d+)}/g, (_, group) => args[group]);
+}
+
+export function mapDistinct(pairs: [string, string][]): Map<string, string> {
+    const distinct = new Map<string, string | undefined>();
+    for (const [key, value] of pairs) {
+        if (distinct.has(key)) {
+            const otherValue = distinct.get(key);
+            if (value !== otherValue) distinct.set(key, undefined);
+        } else {
+            distinct.set(key, value);
+        }
+    }
+    for (const [key, value] of distinct) {
+        if (!value) distinct.delete(key);
+    }
+    return distinct as Map<string, string>;
+}
+
+export function augmentLog(log: Log) {
+    if (log._augmented) return;
+    log._augmented = true;
```

Comment:
This is actually more to the point I made in the original PR. I still believe it would be nice to have the "augmentations" under an object property of LOG:

    interface Log {
        augmentation? {
          _uri: string
          _uriUpgraded?: string // Only present if upgraded.
          _jsonMap?: JsonMap // Only used by the "extension" side for navigating original SARIF sources. The "panel" side does not need this feature and thus does not use this field.
          _augmented: boolean
          _distinct: Map<string, string> // Technically per Run, practically doesn't matter right now.
        }
    }

and this becomes if (log.augmentation) return

and the other code readers do not have to understand the difference between an "_" being present or not.

I'm curious about what Sherry's thoughts are here.

Reply:
Discussed over chat. These underscore prefixes will eventually be removed when we exit preview status.

---
In reply to: [441658562](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441658562) [](ancestors = 441658562)